src/Utils.vala

```diff
@@ -427,45 +438,189 @@ public class FeedReader.Utils : GLib.Object {
+	public static bool downloadFavIcon(string feed_id, string feed_url, string icon_path = GLib.Environment.get_user_data_dir() + "/feedreader/data/feed_icons/")
 	{
-		var path = GLib.File.new_for_path(icon_path);
-		try{path.make_directory_with_parents();}catch(GLib.Error e){}
-		string local_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".ico";
 
-		string url = feed_url;
-		if(feed_url.has_prefix("http://"))
-			url = url.replace("http://", "");
-		else if(feed_url.has_prefix("https://"))
-			url = url.replace("https://", "");
+		var uri = new Soup.URI(feed_url);
+		string hostname = uri.get_host();
+		int first = hostname.index_of_char('.', 0);
+		int second = hostname.index_of_char('.', first+1);
+		if(second != -1 && first != second)
+			hostname = hostname.substring(first+1);
+		string siteURL = uri.get_scheme() + "://" + hostname;
+
+		// download html and parse to find location of favicon
+		var message_html = new Soup.Message("GET", siteURL);
+		if(Settings.tweaks().get_boolean("do-not-track"))
+			message_html.request_headers.append("DNT", "1");
+		var status = getSession().send_message(message_html);
+		if(status == 200)
+		{
+			var html = (string)message_html.response_body.flatten().data;
+
+			var html_cntx = new Html.ParserCtxt();
+	        html_cntx.use_options(Html.ParserOption.NOERROR + Html.ParserOption.NOWARNING);
+	        Html.Doc* doc = html_cntx.read_doc(html, siteURL, null, Html.ParserOption.NOERROR + Html.ParserOption.NOWARNING);
+	        if(doc == null)
+	        {
+	            Logger.debug("Utils.downloadFavIcon: parsing html failed");
+	    		return false;
+	    	}
+
+			// check for <link rel="apple-touch-icon">
+			var xpath = grabberUtils.getURL(doc, "//link[@rel='apple-touch-icon']");
 
-		if(!FileUtils.test(local_filename, GLib.FileTest.EXISTS))
+			if(xpath == null)
+			// check for <link rel="shortcut icon">
+				xpath = grabberUtils.getURL(doc, "//link[@rel='shortcut icon']");
+
+			if(xpath == null)
+				// check for <link rel="icon">
+				xpath = grabberUtils.getURL(doc, "//link[@rel='icon']");
+
+			if(xpath != null)
+			{
+				Logger.debug(@"Utils.downloadFavIcon: xpath success $xpath");
+				xpath = grabberUtils.completeURL(xpath, siteURL);
+				if(downloadIcon(feed_id, xpath, icon_path))
+					return true;
+			}
+			else
+			{
+				Logger.debug("Utils.downloadFavIcon: xpath failed");
+			}
+
+			delete doc;
+		}
+
+		// try domainname/favicon.ico
+		var icon_url = siteURL;
+		if(!icon_url.has_suffix("/"))
+			icon_url += "/";
+		icon_url += "favicon.ico";
+		if(downloadIcon(feed_id, icon_url, icon_path))
+			return true;
+
+
+		// last chance: try allesedv service
+		string allesedv_url = siteURL;
+		if(allesedv_url.has_prefix("http://"))
+			allesedv_url = allesedv_url.replace("http://", "");
+		else if(allesedv_url.has_prefix("https://"))
+			allesedv_url = allesedv_url.replace("https://", "");
+		allesedv_url = "https://f1.allesedv.com/32/%s".printf(allesedv_url);
```

Comment:
I'm not sure if this provides much value if we're already doing the other steps.